azure-iot-device/azure/iot/device/common/mqtt_transport.py

```diff
@@ -353,6 +353,19 @@ def _create_ssl_context(self):
 
         return ssl_context
 
+    def shutdown(self):
+        """Shut down the transport. This is (currently) irreversible."""
+        # Remove the disconnect handler from Paho. We don't want to trigger any events in response
+        # to the shutdown and confuse the higher level layers of code. Just end it.
+        self._mqtt_client.on_disconnect = None
+        # Now disconnect and do some additional cleanup.
+        self._force_transport_disconnect_and_cleanup()
+        # Paho can't be fully shut down until it is garbage collected. So we delete the reference.
+        # Due to the GC algorithm, we don't know exactly when it will get garbage collected, and
+        # thus we don't know when it will truly be shut down. But we do know it will happen (soon).
+        del self._mqtt_client
```

Comment:
>del self._mqtt_client [](start = 8, length = 21)

Do you know with certainty that this is required?  All of the code in this file assumes that `self._mqtt_client` is valid.  It seems like you're taking a big risk in assuming that none of the code in this module is going to be called after `shutdown()` is called.

Reply:
Paho cleans up it's threads in the __del__ method, which is called upon garbage collection. This would automatically happen if our client went out of scope of course (with the IoTHubClient gc'd, then Paho would be as well), but if we want it to happen in response to .shutdown() it would have to be manually deleted.

The other concern is that if we wait for gc, then as long as the IoTHubClient is around (let's say shut down, but not out of scope yet), Paho is still running.

My opinion is that we should delete Paho upon shutdown. If we wanted to create some kind of way to restore the client, we would have to run an InitializePipelineOperation again anyway due to needing to reinitialize pipeline resources like SAS timers etc., and we would create a new instance of the transport as well.

But yeah, honestly the whole thing is kinda messy and has created a lot of edge cases. Perhaps there should be some additional logic added to lock down all other operations when a Shutdown is in progress?

Reply:
I'm not worried about needing to restore the client.   My concern is more about risk.

Don't forget, you're doing this change (the whole PR) because we know that we need to clean up the sas renewal thread.  We knew that this was an example where GC wasn't working.  ie, functionality is broken.   I asked "do you know if this is required" above because it seems like you're cleaning up Paho because "it seems like a good idea" rather than "functionality is broken".  The purpose of shutdown is to "clean up things that we _must_ clean up", not to "clean up everything we can".  We're trying to work around a problem with GC, not "leave everything as clean as possible when we're done."  This is a slippery slope.  

The reason I think this is risky is because of Paho callbacks.  Every time we call into Paho, there's a chance that it will cause Paho to call back into us.  For example (I don't think this can actually happen, but it's an example): 
1. cleanup calls del self._mqtt_client.  The _mqtt_client attribute is removed from our MqttTransport object.  
2. this causes Paho to be collected
3. while being collected, Paho needs to disconnect (I'm not sure if this can happen or not, but theoretically it could.)
4. As part of paho's disconnect logic, it needs to notify us, so it calls mqtt_transport.on_disconnect
5. mqtt_transport.on_disconnect calls _force_transport_disconnect_and_cleanup.
6. _force_transport_disconnect_and_cleanup calls self._mqtt_client.disconnect().

Normally, this circular logic isn't a problem, presumably because Paho doesn't disconnect if it's already disconnected.  

But, because we deleted self._mqtt_client, calling self._mqtt_client.disconnect() raises an AttributeError.  Maybe this is harmless and maybe not.  But, we have an AttributeError that we didn't have before.  This is why I'm concerned.  When you add code to delete the attribute, I then have to assume that every use of self._mqtt_client could cause an AttributeError.  I'm just not smart enough to imagine all the twisted ways that Paho might call back into us.  If I wanted to be extra careful, I would maybe add a call to hasattr before every use of self._mqtt_client, and then slip a little further down that slippery slope.

That being said, you asserted that Paho's __del__ method cleans up its threads.  Maybe I'm missing something, but I don't see where that happens.  The __del__ logic closes sockets, but I don't see it stopping any threads.  

On top of this, `del` doesn't necessarily cause the object to be collected.  It just removes the attribute, which removes the reference, and this may or may not cause the object to be collected.  If anyone else has a reference to the _mqtt_client object (including circular references), it doesn't get cleaned up until the GC decides to clean it up.  

> as long as the IoTHubClient is around (let's say shut down, but not out of scope yet), Paho is still running.
If Paho is still running after shutdown, then there's a bug we need to fix.  _force_transport_disconnect_and_cleanup calls loop_stop, and this should stop the background thread.  Am I reading the code wrong or do you mean something else?

Reply:
You know what, yeah, I think I'll cut this out (for now at least). I'll keep the force stop of Paho in, with the _force_transport_disconnect_and_cleanup() call, but I won't delete the object. It's trivial to add it later, but keeping it does greatly increase our risk exposure on an already fairly substantial PR.